### DISCUSSION TITLE
Bug: There was a problem loading formatter: ...\node_modules\eslint\lib\cli-engine\formatters\stylish

### DISCUSSION THREAD

**Discussion by santi100a** [eslint/eslint#17215] [General] [answered]
### Environment

Node version: v19.6.1
npm version: v9.5.0
Local ESLint version: v8.41.0 (Currently used)
Global ESLint version: Not found
Operating System: win32 10.0.19045

### What parser are you using?

TypeScript Parser (@typescript-eslint/parser)

### What did you do?

<details>
<summary>Configuration</summary>

```
{
    "env": {
        "browser": true,
        "commonjs": true,
        "es2021": true,
        "node": true
    },
    "extends": [
        "eslint:recommended",
        "plugin:@typescript-eslint/recommended"
    ],
    "overrides": [
    ],
    "parser": "@typescript-eslint/parser",
    "parserOptions": {
        "ecmaVersion": "latest"
    },
    "plugins": [
        "@typescript-eslint"
    ],
    "rules": {
    }
}

```
</details>

```typescript
import { deepEquality } from '@santi100/equal-lib';

function indexOf(arr: any[], item: any) {
	for (const idx in arr) {
		if (deepEquality(arr[idx], item)) return arr[idx];
	}
	return -1;
}

export class AssertionError<E = unknown, A = unknown>
	extends Error
	implements AssertOptionalParams<E, A>
{
	/**
	 * Expected value for the assertion.
	 */
	readonly expected: E;
	/**
	 * Received value for the assertion.
	 */
	readonly actual: A;
	/**
	 * Optional operator used for the assertion.
	 */
	readonly operator?: string;
	constructor(expected: E, actual: A, operator?: string) {
		super(
			`Assertion failed! Expected ${expected}. Got ${actual} when using operator ${operator}.`
		);
		this.expected = expected;
		this.actual = actual;
		this.operator = operator;
	}
}
/**
 * Asserts that `condition` is truthy.
 * Throws an {@link AssertionError} otherwise.
 *
 * @param {boolean} condition A boolean condition to assert.
 * @param {AssertOptionalParams} errorParams Parameters for the {@link AssertionError} that will be
 * thrown in case the condition is falsy.
 * @param {unknown} errorParams.expected Expected value for the assertion.
 * @param {unknown} errorParams.actual   Received value for the assertion.
 * @param {string?} errorParams.operator Optional operator used for the assertion.
 * @returns {void} Nothing.
 */
export function assert(
	condition: boolean,
	errorParams: AssertOptionalParams<true, boolean> = {
		expected: true,
		actual: false,
		operator: '==='
	}
): void {
	const { expected, actual, operator } = errorParams;
	if (!condition) throw new AssertionError(expected, actual, operator);
}
/**
 * Asserts that the type of `val` is `expectedType`.
 * Throws an {@link AssertionError} otherwise.
 *
 * @param val An expression whose type is to be asserted.
 * @param expectedType The expected type. Must be one of {@link Type}.
 * @deprecated Use {@link assertTypeOf} instead.
 */
export function assertType<T = unknown>(val: T, expectedType: Type): void {
	const wrongType = typeof val !== expectedType;
	if (wrongType) throw new AssertionError(expectedType, typeof val, 'typeof');
}

export type Type =
	| 'string'
	| 'number'
	| 'bigint'
	| 'boolean'
	| 'symbol'
	| 'undefined'
	| 'object'
	| 'function';
/**
 * The shape of the assertion options.
 */
export interface AssertOptionalParams<E, A> {
	/**
	 * Expected value for the assertion.
	 */
	readonly expected: E;
	/**
	 * Received value for the assertion.
	 */
	readonly actual: A;
	/**
	 * Optional operator used for the assertion.
	 */
	readonly operator?: string;
}
/**
 * Asserts that the type of `arg` is `expectedType`. Throws a `TypeError` otherwise.
 *
 * @param arg An expression whose type is to be asserted.
 * @param expectedType The expected type. Must be one of {@link Type}.
 * @param name An optional expression name to be put in the `TypeError`'s message. Defaults to "arg".
 */
export function assertTypeOf(arg: any, expectedType: Type, name = 'arg') {
	const TYPES = [
		'string',
		'number',
		'bigint',
		'boolean',
		'symbol',
		'undefined',
		'object',
		'function'
	];
	assertOneOf(expectedType, 'expectedType', TYPES);
	if (typeof arg !== expectedType)
		throw new TypeError(
			`"${name}" must be of type "${expectedType}". Got "${arg}" of type "${typeof arg}".`
		);
}
/**
 * Asserts `arg` is one of `choices`.  Throws a `TypeError` otherwise.
 *
 * @param arg The value that's expected to be included within `choices`.
 * @param name An expression name to be put in the `TypeError`'s message.
 * @param choices An array containing the posible values `arg` should have in order for an error not
 * to be thrown.
 * @param shallow Whether or not to use shallow equality (default deep equality is powered by 
   `@santi100/equal-lib` 😉).
 */
export function assertOneOf<T = unknown>(arg: any, name: string, choices: T[], shallow = false) {
	if (shallow ? choices.indexOf(arg) : indexOf(choices, arg) === -1)
		throw new TypeError(
			`"${name}" must be one of "${choices.join(
				', '
			)}". Got "${arg}" of type "${typeof arg}".`
		);
}
function __isInteger(n: number) {
	return n - Math.floor(n) <= 0;
}
/**
 * Asserts `arg` is an integer. Throws a `TypeError` otherwise.
 *
 * @param arg Any number.
 * @param name An optional expression name to be put in the `TypeError`'s message. Defaults to "arg".
 */
export function assertInteger(arg: number, name = 'arg') {
	if (!__isInteger(arg))
		throw new TypeError(
			`"${name}" must be an integer. Got "${arg}" of type "${typeof arg}".`
		);
}
/**
 * Asserts `arg` is bigger than or equal than `min`. Throws a `RangeError` otherwise.
 *
 * @param arg Any value.
 * @param name An expression name to be put in the `TypeError`'s message.
 * @param min The minimum value for `arg`.
 */
export function assertMin(arg: any, name: string, min: any) {
	if (arg < min)
		throw new RangeError(
			`"${name}" must be bigger than or equal to ${min}. Got "${arg}" of type "${typeof arg}".`
		);
}

/**
 * Asserts `arg` is smaller or equal than `max`. Throws a `RangeError` otherwise.
 *
 * @param arg Any value.
 * @param name An expression name to be put in the `TypeError`'s message.

 * @param max The maximum value for `arg`.
 */
export function assertMax(arg: any, name: string, max: any) {
	if (arg > max)
		throw new RangeError(
			`"${name}" must be smaller than or equal to ${max}. Got "${arg}" of type "${typeof arg}".`
		);
}
/**
 * Asserts `arg` is between `min + 1` and `max + 1` (inclusive). Throws a `RangeError` otherwise.
 *
 * @param arg Any value.
 * @param name An expression name to be put in the `TypeError`'s message. 
 * @param min The minimum value for `arg`.
 * @param max The maximum value for `arg`.
 */
export function assertRange(arg: any, name: string, min: any, max: any) {
	if (arg > max || arg < min)
		throw new RangeError(
			`"${name}" must be smaller than ${max} and bigger than ${min}. Got "${arg}" of type "${typeof arg}".`
		);
}
/**
 * Asserts `arg` is an Array. Throws a `TypeError` otherwise.
 *
 * @param arg Any value.
 * @param name An optional expression name to be put in the `TypeError`'s message. Defaults to "arg".
 */
export function assertArray(arg: any, name = 'arg') {
	if (!(arg instanceof Array))
		throw new TypeError(
			`"${name}" must be an Array. Got "${arg}" of type "${typeof arg}".`
		);
}
/**
 * Asserts `arg` is an instance of `clas`. Throws a `TypeError` otherwise.
 * 
 * @param arg An object whose class is to be asserted to be `clas`.
 * @param clas Any valid constructor.
 * @param name An optional expression name to be put in the `TypeError`'s message. Defaults to "arg".
 * @since 1.0.8
 */
export function assertInstanceOf<T = any>(
	arg: any,
	clas: new (...args: any[]) => T,
	name = 'arg'
) {
	try {
		new clas();
	} catch (_) {
		throw new TypeError(
			`"clas" must be a valid constructor. Got ${clas} of type ${typeof clas}.`
		);
	}
	if (!(arg instanceof clas))
		throw new TypeError(
			`${name} must be an instance of ${
				clas.name
			}. Got ${arg} of type ${typeof arg}.`
		);
}
/**
 * Asserts `arg` is smaller than `max`. Throws a `TypeError` otherwise.
 * 
 * @param arg The value whose value is to be asserted to be smaller than `max`.
 * @param max The maximum value `arg` is allowed to have.
 * @param name An optional expression name to be put in the `TypeError`'s message. Defaults to "arg".
 * @since 1.0.8
 */
export function assertExclusiveMax(arg: any, max: any, name = 'arg') {
	if (arg >= max)
		throw new TypeError(
			`"${name}" must be smaller than ${max}. Got ${arg} of type ${typeof arg}.`
		);
}
/**
 * Asserts `arg` is bigger than `min`. Throws a `TypeError` otherwise.
 * 
 * @param arg The value whose value is to be asserted to be bigger than `max`.
 * @param min The minimum value `arg` is allowed to have.
 * @param name An optional expression name to be put in the `TypeError`'s message. Defaults to "arg".
 * @since 1.0.8
 */
export function assertExclusiveMin(arg: any, min: any, name = 'arg') {
	if (arg <= min)
		throw new TypeError(
			`"${name}" must be bigger than ${min}. Got ${arg} of type ${typeof arg}.`
		);
}
```

### What did you expect to happen?

I expected ESLint to run correctly and lint through my source code. 

### What actually happened?

I got this error:

```
There was a problem loading formatter: ...\node_modules\eslint\lib\cli-engine\formatters\stylish
Error: require() of ES Module ...\node_modules\strip-ansi\index.js from ...\assertion-lib\node_modules\eslint\lib\cli-engine\formatters\stylish.js not supported.
Instead change the require of index.js in ...\node_modules\eslint\lib\cli-engine\formatters\stylish.js to a dynamic import() which is available in all CommonJS modules.
```

### Link to Minimal Reproducible Example

https://github.com/santi100a/assertion-lib

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by Rec0iL99** [comment, marked as answer]
Hi @santi100a, thanks for the issue. I wasn't able to reproduce the error you are getting. 

Here are the steps I followed (with output):

```bash
> yarn install

# script from package.json
> yarn lint 

/Users/joelmathew/Desktop/assertion-lib/src/index.ts
    3:23  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
    3:36  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  104:35  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  131:47  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  161:32  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  161:56  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  176:32  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  176:56  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  190:34  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  190:58  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  190:68  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  202:34  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  216:38  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  217:7   warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  218:22  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  243:41  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  243:51  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  257:41  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  257:51  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any

✖ 19 problems (0 errors, 19 warnings)
```

ESLint is working as expected.

It could be an issue with your `node_modules`, I think deleting your `node_modules` and doing a clean install should work .

Thanks for using ESLint.

**Reply by kamilJ96** [reply]
~Getting the same issue after updating to 8.41.0 - have removed `node_modules` and done a clean install multiple times & same thing (same issue in CI with a completely fresh environment)~

edit: This appears to be some weird dependency-hell related issue with some other package

Any package that installs `glob@10.2.2` or above will install the `strip-ansi-cjs` package which is causing this issue.

Add a resolution dependency to pin whatever package this is to the previous one before it installed glob v10 and it should resolve.

(in my case this was `cacache v17.0.6` which was installing glob 10)

**Reply by santi100a** [reply]
Clean install solved the problem!

**Reply by rwam** [reply]
Thanks @kamilJ96, using `glob@10.2.2` and have the same issue: clean install does not solve it. Instead of adding a resolution dependency switching to [fast-glob](https://www.npmjs.com/package/fast-glob) by `yarn remove glob && yarn add fast-glob` fix it to me.

**Reply by thr3a** [reply]
I couldn't solve it with a clean install either.
I was able to resolve the error by explicitly installing the libraries with `yarn add glob@9.3.1` and `yarn add strip-ansi@7.1.0`.

- glob 9.3.1
- strip-ansi 7.1.0
- eslint 8.43.0

**Reply by zoontek** [reply]
Is the maintainer of `glob` aware of this issue?

**Reply by isaacs** [reply]
I am aware. This is a yarn bug. Use npm.

**Reply by guizmaii** [reply]
Migrated from `yarn` to `pnpm` and it fixed the issue 🤷‍♂️

**Reply by rcreasi** [reply]
For me it was sufficient to pin `strip-ansi` to an old version by adding
```
  "resolutions": {
    "strip-ansi": "^6.0.1"
  }
```
to my package.json. In my case there were a few more packages I had to do this with but YMMV.
```
  "resolutions": {
    "strip-ansi": "^6.0.1",
    "string-width": "^4.2.2",
    "wrap-ansi": "^7.0.0"
  }
```

**Reply by anjantalatam** [reply]
npm works.

just do `npm i` and remove yarn.lock for now.

Maybe migrate back to yarn if fixed. ( [Track the issue here](https://github.com/yarnpkg/yarn/issues/8976) )

**Reply by sbaechler** [reply]
Removing yarn.lock and all node_modules folders, then running `yarn install` fixed it for me.

**Reply by adampotency** [reply]
yes but i've had to do it like 20 times and it takes an epic amount of time

**Reply by andreymadnesspartners** [reply]
@sbaechler works once, but appears again after some time

**Reply by Tomekmularczyk** [reply]
yeah, it reappears when you install/update other packages.

**Reply by guilbaultcarbo** [reply]
I have the same issue. The workaround works but then it happens again.

**Reply by karachungen** [reply]
As a workaroud you could put these lines to package.json
```
"resolutions": {
    "strip-ansi": "6.0.1",
    "string-width": "4.2.2",
    "wrap-ansi": "7.0.0"
  },
```

**Reply by elser** [reply]
Thanks. Worth to mention, that after applying this fix you need to remove node_modules once again

**Reply by pcoyoy8** [reply]
For me adding
```
"resolutions": {
    "strip-ansi": "6.0.1"
  },
```
was enough.


Btw, my issue was generated by "eslint" as it cannot use strip ansi 7. Ref: https://github.com/eslint/eslint/issues/15218

**Reply by Frexuz** [reply]
thx @pcoyoy8 this worked for me too :)

**Comment by troyere** [comment]
Got the error too, when using **latest versions** at this day :

```
There was a problem loading formatter: [...]/node_modules/eslint/lib/cli-engine/formatters/stylish
Error: require() of ES Module [...]/node_modules/strip-ansi/index.js from [...]/node_modules/eslint/lib/cli-engine/formatters/stylish.js not supported.
Instead change the require of index.js in [...]/node_modules/eslint/lib/cli-engine/formatters/stylish.js to a dynamic import() which is available in all CommonJS modules.
error Command failed with exit code 2.
```

**Reply by mhp-borisbojic** [reply]
Happens to me as well. Adding the resolutions mentioned above, helps, but this can't be the long term solution.

**Reply by troyere** [reply]
I refuse to do that lol. Instead, I gave up using `stylish` everywhere : I didn't choose it, and its full of bugs. It ends now. 

I use the [visualstudio](https://eslint.org/docs/latest/use/formatters/) eslint formatter, it works well.

**Reply by Idildndr** [reply]
what did you use instead of stylish, my error is from that line.                         formatter: require('eslint').CLIEngine.getFormatter('stylish')

**Comment by johnnyshankman** [comment]
Having this too. Dang. What works in one environment wont work in another using the same version of node.

```
"resolutions": {
    "strip-ansi": "6.0.1"
}
  ```
  
Worked which is absolutely not a good solve. Also breaks once you intstall a new pkg without regenerating the lock.

Why is this affecting yarn repos? Can we get more information on the actual bug here? I can't get anything to work on v18 unless I do this everywhere. Willing to lend a hand with a PR.

**Update: long term the best solve really is to just use npm.**

**Reply by marcel-kuehn** [reply]
I have the same issue, also using yarn.

**Reply by isaacs** [reply]
Why are you hitting yourself? Stop hitting yourself.

Seriously. The yarn version you're insisting on using has this bug, and you're complaining about the bug. The bug won't be fixed. You know this, it's been stated everywhere this issue comes up. Use a package manager that works, or just accept that you're inflicting this pain on yourself.

npm has supported this feature for many years. It's included with node already, you don't even have to go get it. So has pnpm. Even yarn 2 supports it. Yarn 1 is broken, and will never be fixed. If you want more info on why yarn has this bug, ask yarn, not the packages yarn fails to install properly.

In other news: I can't get eslint to work on cobol running on my VAX/VMS mainframe, and when I bang rocks together it doesn't summon Santa Claus. Can someone fix these or give me more info on them? It's so sad!

**Reply by marcel-kuehn** [reply]
Uhm, what are you talking about, I had this bug today and came across this issue. I have nothing to do with the person commenting above. I am also not using yarn 1 btw. No reason to be so rude.

**Reply by johnnyshankman** [reply]
hate to be the bearer of bad news to you @isaacs , but yarn 1 ("yarn classic") is the most widely used version of yarn, more than yarn 2 or yarn 3 ("yarn modern"). not that it's better, it isn't, but it's considered a stable classic version and since it works just fine for every other package besides eslint in this edge case, it's a great off-the-shelf choice, as dev ops never need to worry about it updating and it's supported on almost every deployment system and machine around (Github Actions, Vercel, Fleek, etc). been the case since 2015.

i also think pointing the finger at yarn here is a bit uncalled for. i have this issue with 0 other packages and use yarn classic as a daily driver, have been for 5 years. this repo is plagued by an edge case issue they patched in yarn 2, but we can also just fix the core issue ourselves in this repo instead of pushing users to yarn modern. it's a very reasonable ask for a widely used package such as eslint.

like i mentioned above, happy to provide a PR if anyone can actually provide a push in the right direction.

**Reply by isaacs** [reply]
> but it's considered a stable classic version and since it works just fine for every other package besides eslint in this edge case,

As the author and maintainer of several hundred npm packages, many of which are used in virtually every single JavaScript dev tool, I can say with the greatest confidence possible that it does not, in fact, work "just fine for every other package besides eslint in this edge case". Anything that uses glob, rimraf, or countless others will encounter this at some point, and are only working by accident if they are. And I'm not the only author so afflicted with this behavior.

Any time a package uses `npm:` dependency aliases (introduced in 2018, fwiw), if the tree ends up with two different actual versions of the source, yarn 1 is liable to serve packages the incorrect versions of their dependencies.

Now, that might not _break_ your program, but it is _broken behavior_ that happens very regularly. Every time it happens, maybe 99% of people don't encounter the problem, since "breaking changes" often aren't actually "breaking" in the conventional sense. But that 1% is hundreds or thousands of people, and many of them get very upset when the complain to me about my code not working in their project, and I (accurately!) point the finger at their broken, unsupported, unmaintained, and _unstable_ package manager which is the culprit.

I cannot begin to explain to you how frustrating this is, as a package author, the author of npm, and original developer of node's module system _all specifically designed to make it possible to avoid exactly this error_, having done all of this work in the open since 2010, having literally begged yarn to just take it and use it, to now repeatedly be told that this clear and obviously broken behavior on yarn's part (which I have been complaining about since 2015, as it was broken on day one, at has never worked properly) simply _must_ be something else, it's unfair to blame yarn, clearly my code has to change to accommodate it, etc.

No. You're mistaken. Yarn is broken. You're not wrong to use it, but you're definitely wrong to believe it's anyone else's problem if you do so. I am not going to do any more to point you in the right direction, because the "direction" you need to be pointed in is _away from yarn 1_, since it's broken, and will never be fixed. I don't know how to make it any clearer. I wrote a whole package manager for you that doesn't have this bug. 2 other projects (pnpm and yarn post-v1) have as well. I'm not going to accommodate broken package managers when I already gave you a working one for free.

**Reply by isaacs** [reply]
> like i mentioned above, happy to provide a PR if anyone can actually provide a push in the right direction.

The PR would be to yarn. It won't be accepted. Yarn 1 is not supported. It has been abandoned, and you would be wise to abandon it as well.

**Reply by isaacs** [reply]
@marcel-kuehn If you aren't using yarn 1, and aren't getting the incorrect versions of things served by your package manager, then you don't have "the same bug" you have a different bug that looks similar.

**Reply by johnnyshankman** [reply]
you're correct and i understand the issue. would be nice to have this documented in the readme since it's something you're clearly *very aware* of. no need for the anger. we're all just trying to help. if anger is your reaction to bug reports that are considered closed but still getting this many active comments (and the answer is completely useless and does not contain a solve nor help), i suggest you take a breather from working on this repo.

**Reply by isaacs** [reply]
@johnnyshankman I'm not angry at anyone here. I know tone can be hard to read in the best of times, all the more so in text. But really, my emphatic tone here is just frustration with the fact that so many insist on inflicting unnecessary pain on themselves and don't see it.

To the extent that there is any anger, it's towards the original authors of yarn who refused to consider the impact of their NIH-ism, and Facebook's marketing department who spent so much effort telling lies about a project I spent so much time on, and then abandoning the project once it turned out that they really couldn't leverage it to increase Facebook's influence over the OSS community. Definitely still a bit salty about that, probably always will be 😅 We give too much to giant corporations, and almost never hold them accountable when they ruin things for the rest of us.

None of that is aimed at the current team who wrote yarn 2, especially Mael, who seems a decent and very conscientious OSS developer.

**Reply by arcanis** [reply]
I'm going to be so decent and conscientious I'm also going to pretend this wall of text from an npm author doesn't exist. It would be beyond pale if it did. What I'm going to do however, is [ask for a reproduction somewhere down this thread](https://github.com/eslint/eslint/discussions/17215#discussioncomment-8072242) from someone who cares more about collaboration than trashtalk.

**Reply by isaacs** [reply]
@arcanis Don't be passive aggressive, it's undignified. If I said something untrue, please, show me where I'm mistaken.

Yarn 1 is clearly doing the wrong thing here, and I stand by my assessment that it should not be used. If it took me 5 minutes to reproduce it by just trying the most obvious steps to put the conflicting packages in the same project, then I don't believe that you actually tried, because you surely would have found it just as quickly.

I'm not just "an npm author", I'm also a package maintainer, and I get yarn users in my issues constantly due to this. It's a serious shortcoming, and it seems no one is bothering to even look at it. Which, actually, is completely fine, and I don't take issue with at all. "THE SOFTWARE IS PROVIDED AS IS" is what makes this whole OSS thing work.

But if that's the case, and the intent is for yarn 1 to be unsupported, then the responsible thing to do is fully deprecate yarn 1. That's the part that I find annoying: that yarn classic is presented as "stable", when it is in fact pretty badly broken, always has been, and there are no plans to fix it.

**Reply by arcanis** [reply]
Only thing I'll bother to say is that as a package maintainer I'd expect you to know that _"It seems to be the `npm:` hoisting issue; @arcanis, are you aware of it?"_ is better received than the typical pre-2020 npm tweetstorm dramas I absolutely didn't miss.

**Reply by isaacs** [reply]
@arcanis It would be more pleasant for you if I was more pleasant, I'm sure. But I disagree that it would be _better_ received. As far as I can tell, it was not received at all until I shared this frustration and resentment. 🤷

**Reply by Johnrobmiller** [reply]
@arcanis When someone says mean things about the project you're working on, getting emotional and defensive is a bad move. Better to just fully accept the problem, which means fully accepting the criticisms about the problem. This is an ideal attitude for fixing problems and not making excuses.

All of this fits under the overarching "umbrella value" that is accountability. When you fully accept criticism, you put yourself in a position to be accountable for it.

**Reply by isaacs** [reply]
Maybe it's the tism talking, but I really don't even see what was mean about what I had to say about yarn? Yarn 1 has a bug that leads to serving incorrect dependency versions. That bug has been around for many years, has been brought up a bunch of times, and isn't being fixed. That's 100% fine, "AS IS" etc. But likewise, I'm not responsible for making _my_ code work with a tool that has such a critical known bug and isn't being fixed. People ask me to fix this bug, but it's a bug in a project I don't own or really know all that much about, so yeah, I'm gonna say "don't use the broken thing", but then people get mad. Hence: why are you hitting yourself? Stop that.

Lots of my code has bugs, I'd be willing to guess almost all of the code I've ever written has had bugs that had to be fixed, and every project eventually has versions that get abandoned and will never be fixed, that's just how it goes. Is that a mean thing to say? It seems pretty normal and obvious, I'd think.

**Reply by heyakyra** [reply]
Just to connect people who end up at either of these from a search engine, so they can get answers:
* https://github.com/eslint/eslint/issues/13043
* https://github.com/eslint/eslint/discussions/13949

**Comment by mdjermanovic** [comment]
To clarify, ESLint requires strip-ansi v6:

https://github.com/eslint/eslint/blob/d31c180312260d1a286cc8162907b6a33368edc9/package.json#L99

By the error messages posted in this discussion, it seems that ESLint is getting strip-ansi v7 to work with.

I'm not sure what exactly is causing this, but this is not a problem in ESLint. It's the responsibility of package managers to provide packages with dependencies that satisfy the declared ranges. `7.x` does not satisfy `^6.0.1`.

**Reply by isaacs** [reply]
Yes, this is 100% a yarn bug, and it's a common one I get to hear about very frequently. https://github.com/eslint/eslint/discussions/17215#discussioncomment-8042361

**Comment by johnnyshankman** [comment]
~~@isaacs~~ we should mark this package doesn't work with yarn 1. this just isn't clearly documented, so clearly we're going to have tons of people pouring in reporting the issue.

-------

![Screenshot 2024-01-08 at 10 31 20 AM](https://github.com/eslint/eslint/assets/6632701/bd23d85d-e371-4b83-bf8c-b9c37a7e4d67)

---------- 

**just let us know ahead of time so we don't get confused and open an issue!**

> Anything that uses glob, rimraf, or countless others will encounter this at some point...

Also the thing marked as the "answer" just says it cant recreate. Let's mark one of the various hacks as the solve for v1 yarn users and attach an explainer!

**Reply by platinumazure** [reply]
Isaacs is not a maintainer of ESLint. Please don't ask him to do anything more than he is already doing, which is volunteering his time to explain why an unsupported package manager is broken. He's already doing much more than could be expected.

I can't speak to whether the team will or should add a note about Yarn 1 being unsupported.

**Reply by isaacs** [reply]
Thanks for the understanding words, @platinumazure 

Yeah, even if I did add a warning in glob (really, even deeper than that, in [jackspeak](http://npm.im/jackspeak)) no one installing ESLint (or any of the other tens of thousands of glob dependents) would be likely to see it, since it'd be in a readme 2 layers deeper in the dep tree than they're likely even aware of.

There's probably some way to unambiguously determine whether it's yarn 1 that's doing the installation, and crash or something. I've considered it, believe me. But the hazard there is that _it does work sometimes_, as long as nothing in your tree depends on getting the correct version of any of the aliased dependencies, so it feels a bit too heavy handed to take that approach.

**Reply by johnnyshankman** [reply]
oh i had no idea my bad, @isaacs thank you so much for your help this far and @platinumazure for the explainer. 

highly grateful for it. you're a legend in the OSS community and we're lucky to have your brainpower for problems like these <3

**Comment by arcanis** [comment]
Would someone have an actual reproduction to share? I'm curious about that, but I can't find one in this thread or the [Yarn issue](https://github.com/yarnpkg/yarn/issues/8976), so it's difficult to find out exactly what happens and understand whether this is expected or not. Help appreciated!

**Reply by isaacs** [reply]
This happens when yarn v1 is used to install a package (such as `@isaacs/cliui`) that uses `npm:` aliases, where another package in the tree depends on the aliased version and will break if it gets the wrong version. Yarn v1 serves it the wrong version, which cannot be loaded in this case, because one is ESM and the other is CJS.

This is the set of deps that trigger the issue:

```
$ npm view @isaacs/cliui dependencies
{
  'string-width': '^5.1.2',
  'string-width-cjs': 'npm:string-width@^4.2.0',
  'strip-ansi': '^7.0.1',
  'strip-ansi-cjs': 'npm:strip-ansi@^6.0.1',
  'wrap-ansi': '^8.1.0',
  'wrap-ansi-cjs': 'npm:wrap-ansi@^7.0.0'
}
```

Run these commands in a new folder:

```bash
echo {} > package.json
npx yarn@1 add @isaacs/cliui
npx yarn@1 add eslint
npm ls -a
```

Note the multiple errors.

The package at `./node_modules/eslint` depends on getting `strip-ansi` at version 6. But the package at `./node_modules/strip-ansi` is version 7, and there is nothing present at `./node_modules/eslint/node_modules/strip-ansi` to override it.

**Reply by arcanis** [reply]
Interesting, thanks for the reproduction. I have a tentative fix in https://github.com/yarnpkg/yarn/pull/9023, I want to think about it some more before landing it.

**Comment by andrevitalb** [comment]
I was able to solve this issue by just deleting the lock file and reinstalling dependencies

**Reply by kai-phan** [reply]
Me too. This worked like a champ

**Reply by DevJoseManuel** [reply]
The same thing has happened to me.

**Reply by wolfitoXtreme** [reply]
If you add/upgrade/remove another dependency the problem will be there again, so you'll have to repeat the process of deleting and installing manually again, NOT a solution

**Reply by wolfitoXtreme** [reply]
I can confirm that [adding the `strip-ansi` resolution](https://github.com/eslint/eslint/discussions/17215#discussioncomment-7891807) is a more reliable solution appart from upgrading to yarn 4.

**Comment by Rasukarusan** [comment]
I originally used yarn, but after doing a `npm install` to try it out, I no longer get the error.
Then I did `yarn lint` and no more errors.

**Comment by simi** [comment]
I just recreated `yarn.lock` (`rm yarn.lock`, `yarn install`) to fix this problem.

**Reply by dvzrd** [reply]
That's a temp solution, it will come up again after you add, remove or upgrade deps.

**Comment by Doko-Demo-Doa** [comment]
Upgraded to Yarn 4 as suggested. Problem gone. Nothing else need to be changed. Cheers.

**Comment by Frexuz** [comment]
I think a package should work in any yarn version. yarn v1.x is still the most used version and its the default
even more so, in CI environments. Installing different yarns in those environements, or across your team, is not trivial.
Things like `patch-package` also dont work in yarn v2+

**Reply by arcanis** [reply]
> I think a package should work in any yarn version. yarn v1.x is still the most used version and its the default
> even more so, in CI environments. Installing different yarns in those environements, or across your team, is not trivial.
> Things like `patch-package` also dont work in yarn v2+

You don't need to use patch-package in modern Yarn releases, since it supports package patching natively

With that said, a fix for the reported issue landed in a patch 1.22 release already. Make sure you update if you still face it (and open a new issue on the Yarn repo if not; I won't monitor this GH discussion thread).

**Reply by Frexuz** [reply]
I'm already on 1.22.22 🤔

**Comment by paolagd** [comment]
Deleting the lock file and running `yarn install` worked for me. I am using the latest version of yarn and ESLint 9.1.1.

**Comment by katiacih** [comment]
Fixed

1- check you have in eslint config parser: "@babel/eslint-parser"
2- remove yarn.lock
3 - reinstall dependencies

**Reply by Yarchick1034** [reply]
Such manipulations are temporary...
